Make Gaussian Process Libtorch Compatible

DanielYankura opened the pull request:
refs #30564

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
Utilizing Libtorch would add more capabilities to Gaussian Process
## Design
<!--A concise description (design) of the enhancement.-->
Replace all Eigen::Matrices with torch::Tensors
Replace methods that have a Libtorch equivalent.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
This PR alone should not require re-golding of tests. However it should open up possibilities of future feature additions using libtorch.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

DanielYankura committed:
First commit to make Gaussian Process Libtorch compatible

- First all files related to Gaussian Process were duplicated and made part of a new torch action
- This is just for testing purposes. Once all changes have been proven to work they can be brought over to the original files.
- To test with these modified files add Torched to the gaussian process action in your input file
- In GaussianProcessTorched.C several functions have already been modified to work with torch::tensor

refs #30564

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2886880/), step Clang format on 6caa619 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30565/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30565/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e67d425ea0da66e98b07fcb260321313ef9b4315`

DanielYankura committed:
More vectors changed to torch::Tensor(s)

- Implemented basic test for finding mean
- Replaced a number of vectors with torch::Tensor(s) to implement test

DanielYankura committed:
Added more tests for standardizer

- Added equivalent tests for both Eigen and libtorch versions so that runtimes and output can be compared
- Removed outdated comments from previous commit
- Need to make test matrices larger and more complex so that runtimes can be more effectively

DanielYankura committed:
Replacing K Cholesky Decomp with torch::Tensors

-Throughout Gaussian Process Eigen::Matrices were replace with torch::Tensor
-Functions that calculate and solves decomp were replaced with libtorch equivalent
-Made copy of tests that now uses the Libtorch version but check with same gold files
-Until dataStore and dataLoad are torchified not all will pass

DanielYankura committed:
style changes to pass precheck

DanielYankura committed:
Added newline to EOF

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3986356/), step Docs: sync website on 6ff995d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30565/site)

This comment will be updated on new commits.

DanielYankura committed:
Framework for getting covariances converted to libtorch

-Made libtorch versions of each covariance file
-Added changes to framework to support this
-To activate replace Covariance with CovarianceTorched in input file
-Replace covar_function with covar_function_torched
-Now calling this returns a CovarianceFunctionBaseTorched object
-Right now calling this will run the same process as before, but that will be changed in later commits

DanielYankura committed:
Converted all covariance functions to use torch::Tensor

-Covariance function objects now take in torch tensors as constructor params
-All covariance functions were modified to do the same operations but with tensors
-A bug was also fixed that caused 4 tests to error instead of reporting a csv diff
-Now 5 tests pass, 4 report a csv diff, and 1 errors out

DanielYankura committed:
Enabling storing and loading data

-Added store/loadData functions that can be used with torch::Tensor
-getDestandardized test modified to first standardize, then destandardize test matrix
-There is a bug in StandardizerTorched.C that is causing 5 tests to report CSV diffs

DanielYankura committed:
Various fixes

-Fixes made to Standardizer to fix tensor operations
-Now tensors are being returned with the correct dimensions and values
-Fixes to squared exponential covariance

DanielYankura committed:
removed print statements

DanielYankura committed:
Removed last of Eigen matrices from gaussian process class

DanielYankura committed:
Removed last of Eigen matrices from GP Surrogate and GP Trainer

DanielYankura committed:
Moved over all changes from the Libtorch file versions to the original

-File headers and function calls will have to be modified again

DanielYankura committed:
moved all changes from copied src files back to original

DanielYankura committed:
Moved changes from torched src files to original ones

DanielYankura committed:
Replaced Eigen matrices with libtorch tensors in Active Learning GP

DanielYankura committed:
Modified so that getMean and getStdDev calls work with tensors

DanielYankura committed:
Modifications needed to make file name change work

- Fixes to several covariance and standardizer functions that were not being called correctyle before filename change
- Removal of duplicate tests and outdated standardizer unit test
- Fixes to GP were tensor was being reshaped incorrectly
- Fixes to the various set functions in Standardizer
- Fixes to LMC where computeAGradient was not updating the grad tensor properly

DanielYankura committed:
Added Libtorch ifdef to all GP files, and removed unused overload of dataStore/dataLoad

DanielYankura committed:
added libtorch requirement to all gaussian process tests

DanielYankura committed:
added newline to eof

DanielYankura committed:
Added libtorch enabled ifdef to classes that rely of gaussian process

DanielYankura committed:
correct sign type comparison

DanielYankura committed:
More modifications to pass all tests

-Added Libtorch_Enabled ifdef to more classes that relied on gaussian process
-Modified related tests to only run when Libtorch was enabled
-Modified gold file for one example after removing calls to libtorch related functions from its input file

DanielYankura committed:
added type caste for comparison of unsigned ints

DanielYankura committed:
added libtorch requirement to corresponding documentation

DanielYankura committed:
Efficiency changes

- Moved accessor creation outside of for loop
- Replaced data_ptr with torch accessors where possible

DanielYankura committed:
Added ifdef statement surrounding Libtorch include statement

DanielYankura committed:
changed a redefinition of a tensor to a resize

DanielYankura committed:
various efficiency changes

-Moved accessor creation out of for-loop
-In getLoss and getGradient, tensor accessors were replaces with .item<Real> for single element tensors

grmnptr commented on modules/stochastic_tools/src/covariances/ExponentialCovariance.C:
```diff
@@ -40,54 +41,57 @@ ExponentialCovariance::ExponentialCovariance(const InputParameters & parameters)
   for (unsigned int ii = 0; ii < num_samples_x; ++ii)
   {
     for (unsigned int jj = 0; jj < num_samples_xp; ++jj)
     {
       // Compute distance per parameter, scaled by length factor
       Real r_scaled = 0;
       for (unsigned int kk = 0; kk < num_params_x; ++kk)
-        r_scaled += pow((x(ii, kk) - xp(jj, kk)) / length_factor[kk], 2);
+        r_scaled += pow((x_accessor[ii][kk] - xp_accessor[jj][kk]) / length_factor[kk], 2);
       r_scaled = sqrt(r_scaled);
-      K(ii, jj) = sigma_f_squared * std::exp(-pow(r_scaled, gamma));
+      K_accessor[ii][jj] = sigma_f_squared * std::exp(-pow(r_scaled, gamma));
     }
     if (is_self_covariance)
-      K(ii, ii) += sigma_n_squared;
+      K_accessor[ii][ii] += sigma_n_squared;
   }
 }
```
This can be a potential cause of slowdown. Libtorch has optimized routines to compute these matrix operations.

grmnptr commented on modules/stochastic_tools/src/covariances/LMC.C:
```diff
@@ -63,31 +64,32 @@ LMC::LMC(const InputParameters & parameters)
     _covariance_functions[exp_i]->computeCovarianceMatrix(K_params, x, xp, is_self_covariance);
     computeBMatrix(B, exp_i);
-    MathUtils::kron(K_working, B, K_params);
+    // MathUtils::kron(K_working, B, K_params);
```
```suggestion
```

grmnptr commented on modules/stochastic_tools/src/covariances/LMC.C:
```diff
@@ -123,15 +125,16 @@ LMC::computedKdhyper(RealEigenMatrix & dKdhp,
       _covariance_functions[number]->computeCovarianceMatrix(K_params, x, x, true);
     }
-    MathUtils::kron(dKdhp, dBdhp, K_params);
+    // MathUtils::kron(dKdhp, dBdhp, K_params);
```
```suggestion
```

grmnptr commented on modules/stochastic_tools/src/covariances/LMC.C:
```diff
@@ -158,39 +162,45 @@ LMC::computedKdhyper(RealEigenMatrix & dKdhp,
     {
-      Bmat(row_i, col_i) = a_coeffs[row_i] * a_coeffs[col_i];
+      Bmat_accessor[row_i][col_i] = a_coeffs[row_i] * a_coeffs[col_i];
       if (row_i == col_i)
-        Bmat(row_i, col_i) += lambda_coeffs[col_i];
+        Bmat_accessor[row_i][col_i] += lambda_coeffs[col_i];
```
this is just a*a^T+lambda*I, can be defnitely accelerated with matrix operations in torch.

grmnptr commented on modules/stochastic_tools/src/covariances/LMC.C:
```diff
@@ -158,39 +162,45 @@ LMC::computedKdhyper(RealEigenMatrix & dKdhp,
 {
-  grad.setZero();
-  grad(index, index) = 1.0;
+  // grad.setZero();
```
```suggestion
```

grmnptr commented on modules/stochastic_tools/src/covariances/MaternHalfIntCovariance.C:
```diff
@@ -41,33 +42,37 @@ MaternHalfIntCovariance::MaternHalfIntCovariance(const InputParameters & paramet
+  auto xp_accessor = xp.accessor<Real, 2>();
 
-  mooseAssert(num_params_x == xp.cols(),
+  unsigned int num_samples_x = x.sizes()[0];
+  unsigned int num_samples_xp = xp.sizes()[0];
+  unsigned int num_params_x = x.sizes()[1];
```
```suggestion
  const unsigned int num_samples_x = x.sizes()[0];
  const unsigned int num_samples_xp = xp.sizes()[0];
  const unsigned int num_params_x = x.sizes()[1];
```

grmnptr commented on modules/stochastic_tools/src/covariances/SquaredExponentialCovariance.C:
```diff
@@ -73,17 +78,18 @@ SquaredExponentialCovariance::SquaredExponentialFunction(RealEigenMatrix & K,
       for (unsigned int kk = 0; kk < num_params_x; ++kk)
-        r_squared_scaled += std::pow((x(ii, kk) - xp(jj, kk)) / length_factor[kk], 2);
-      K(ii, jj) = sigma_f_squared * std::exp(-r_squared_scaled / 2.0);
+        r_squared_scaled +=
+            std::pow((x_accessor[ii][kk] - xp_accessor[jj][kk]) / length_factor[kk], 2);
+      K_accessor[ii][jj] = sigma_f_squared * std::exp(-r_squared_scaled / 2.0);
```
Yeah there is an optimized way to do this one, might help. You can use ||x-y||^2 = ||x||^2 + ||y||^2 + 2 x*y.

grmnptr commented on modules/stochastic_tools/src/libtorch/surrogates/LibtorchANNSurrogate.C:
```diff
@@ -44,12 +44,18 @@ LibtorchANNSurrogate::evaluate(const std::vector<Real> & x) const
+  auto input_mean_accessor = input_mean.accessor<Real, 2>();
+  auto input_std_accessor = input_std.accessor<Real, 2>();
+
+  // unsigned long int i_mean = (unsigned long int)torch::size(input_mean, 0);
+  // unsigned long int i_std = (unsigned long int)torch::size(input_std, 0);
```
```suggestion
```

zachmprince replied:
Remove

grmnptr commented on modules/stochastic_tools/src/utils/GaussianProcess.C:
```diff
@@ -67,35 +68,39 @@ GaussianProcess::linkCovarianceFunction(CovarianceFunctionBase * covariance_func
-  _K_results_solve = _K_cho_decomp.solve(input);
+  // llt does the llt decomposition as part of the Cholesky module
+  _K_cho_decomp = torch::linalg_cholesky(_K);
+  _K_results_solve = torch::cholesky_solve(input, _K_cho_decomp);
```
I think this will do another cholesky decom of the cholesky (itself) in the background. Would be interesting to see if we could do the two without the double decomposition.

grmnptr commented on modules/stochastic_tools/src/utils/GaussianProcess.C:
```diff
@@ -188,11 +197,17 @@ GaussianProcess::tuneHyperParamsAdam(const RealEigenMatrix & training_params,
-        inputs(ii, jj) = training_params(v_sequence[ii], jj);
+      for (unsigned int jj = 0; jj < training_params.sizes()[1]; ++jj)
+      {
+        // inputs(ii, jj) = training_params(v_sequence[ii], jj);
```
```suggestion
```

grmnptr commented on modules/stochastic_tools/src/utils/GaussianProcess.C:
```diff
@@ -188,11 +197,17 @@ GaussianProcess::tuneHyperParamsAdam(const RealEigenMatrix & training_params,
-        outputs(ii, jj) = training_data(v_sequence[ii], jj);
+      for (unsigned int jj = 0; jj < training_data.sizes()[1]; ++jj)
+      {
+        // outputs(ii, jj) = training_data(v_sequence[ii], jj);
```
```suggestion
```

grmnptr commented on modules/stochastic_tools/src/utils/GaussianProcess.C:
```diff
@@ -263,8 +279,10 @@ GaussianProcess::getGradient(RealEigenMatrix & inputs) const
-      RealEigenMatrix tmp = alpha * dKdhp - _K_cho_decomp.solve(dKdhp);
-      grad_vec[global_index] = -tmp.trace() / 2.0;
+      // This is just temporary. Convert dKhp_matrix to tensor
+      auto tmp = torch::trace(torch::mm(alpha, dKdhp) - torch::cholesky_solve(dKdhp, _K_cho_decomp))
```
This can be expensive if the database is large. We are doing another solve in the background I think.

grmnptr commented on modules/stochastic_tools/src/utils/GaussianProcess.C:
```diff
@@ -263,8 +279,10 @@ GaussianProcess::getGradient(RealEigenMatrix & inputs) const
+      // This is just temporary. Convert dKhp_matrix to tensor
+      auto tmp = torch::trace(torch::mm(alpha, dKdhp) - torch::cholesky_solve(dKdhp, _K_cho_decomp))
+                     .item<Real>();
+      grad_vec[global_index] = -1 * (tmp / 2.0);
```
```suggestion
      grad_vec[global_index] = tmp / -2.0;
```

grmnptr replied:
Would this work?

grmnptr commented on modules/stochastic_tools/test/tests/surrogates/gaussian_process/GP_squared_exponential_torched.i:
```diff
@@ -0,0 +1,123 @@
+[StochasticTools]
```
It think you can remove this at this point.

grmnptr commented on modules/stochastic_tools/src/surrogates/GaussianProcessSurrogate.C:
```diff
@@ -79,34 +80,43 @@ GaussianProcessSurrogate::evaluate(const std::vector<Real> & x,
-      K_test - (K_train_test.transpose() * _gp.getKCholeskyDecomp().solve(K_train_test));
+  torch::Tensor pred_var =
+      K_test - torch::mm(torch::transpose(K_train_test, 0, 1),
+                         torch::cholesky_solve(K_train_test, _gp.getKCholeskyDecomp()));
```
I think this line causes most of the slowdown.

grmnptr reviewed (changes requested):
Super nice! Let's talk about how to make it faster!

DanielYankura committed:
changes for effeciency and formatting

-Replaced element-wise covariance calculation with tensor equivalents
-Replaced cholesky_solve with linalg_solve
-Removed outdated comments
-Removed unused test file

DanielYankura committed:
Removed unused variables and old comments

grmnptr committed:
Update to current libtorch infrastructure.

grmnptr committed:
Optimize the hot loops, to make sure we dont see a lot of slowdown.

grmnptr committed:
Make the current implementation a little safer.

grmnptr committed:
Transition towards tenosrs holding hyperparams as well.

grmnptr committed:
Add simplification to how hyperparams are stored within the covariance objects.

grmnptr committed:
Further simplify the hypper parameter storage.

grmnptr committed:
Add formatting, remove tabs.

grmnptr committed:
Mak eint casts more libmeshy.

grmnptr committed:
Add moarr tensorization.

grmnptr committed:
Clean up adam GP training.

grmnptr committed:
Regold GP example and surrogate regression tests.

grmnptr committed:
Regold active learning and multioutput GP.

grmnptr committed:
Regold the 4D example.

grmnptr committed:
Add utilities to convert tensors from and to device.

grmnptr committed:
Add option to do GP training on GPUs

grmnptr committed:
Add test specs so that GP tests are run on GPU as well.

grmnptr committed:
Make LMC libtorch-dependent in terms of doco.

grmnptr committed:
Clean up dataIO, and make hyperparameter checks static on covariancbase.

grmnptr committed:
Fix active learning no-torch compilation.

grmnptr committed:
Make training with smaller batches simpler.

grmnptr committed:
Regold heavy test too, add some memory reduction and slots to the tests.

grmnptr committed:
Tune active learning cases to be less computational heavy.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3821445/), step Results summary on 8b21dc7 wanted to post the following:

## Framework test summary

Compared against 0ad1875 in job [civet.inl.gov/job/3816166](https://civet.inl.gov/job/3816166).

No change
## Modules test summary

Compared against 0ad1875 in job [civet.inl.gov/job/3816166](https://civet.inl.gov/job/3816166).

### Added tests

| Test                                                                                                | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------------------------------------|------------|---------------|
| `stochastic_tools/test:surrogates/gaussian_process.tuning/GP_squared_exponential_tuned_legacy_adam` | 2.27       | 217.60        |

grmnptr committed:
Adjust slots on parallel active learning.

grmnptr commented:
Okay, I investigated the increase in the memory consumption. Well AI investigate it for me:

The heavy part is PyTorch’s CPU runtime and dependency tree, not the tiny `libtorch.dylib` shim and not the GP matrix data for these tests.

What I measured locally:

Probe / case |	Max RSS
minimal C++ binary linked to libtorch.dylib, does no tensor work | ~86 MiB
same, but does small randn, mm, linalg_cholesky, cholesky_solve, backward | ~95 MiB
stochastic_tools-opt --version | ~146 MiB
ActiveLearningGP COV, num_random_seeds=64	| ~167 MiB
ActiveLearningGP COV, num_random_seeds=10000 | ~217 MiB
ActiveLearningGP COV, default 100000 seeds | ~670 MiB

So the libtorch/tensor part is roughly:

~80-90 MiB RSS fixed per process just from loading the libtorch runtime.
~10 MiB more when small ATen/autograd/linalg tensor paths are first exercised.
The GP tensors in these tests are small; they are not hundreds of MiB.
The concrete heavy library is not libtorch.dylib itself. That file is only ~20 KiB. It loads:

libtorch_cpu.dylib      ~170 MiB on disk
libc10.dylib            ~716 KiB
libomp.dylib            ~828 KiB
libtorch_cpu.dylib is the big one.

A vmmap snapshot of a sleeping no-op libtorch-linked process showed:

Physical footprint: ~49 MiB
ps RSS:             ~76-78 MiB
Default malloc:     ~41 MiB allocated/dirty
libtorch_cpu text:  ~22 MiB resident
libtorch_cpu data:  ~2.6 MiB resident/dirty

So the expensive fixed part is:

1) PyTorch CPU operator/runtime registration
The no-op linked process already has ~41 MiB dirty malloc before any tensor work. I cannot attribute every allocation to a symbol without malloc stack logging, but this probe has no MOOSE and no GP. That memory is from the libtorch dependency tree and static/runtime initialization.

2) libtorch_cpu.dylib code and metadata
The binary maps a ~170 MiB PyTorch CPU library. Not all of that is private or even resident, but enough code/metadata pages become resident to show up in RSS. On macOS, the harness falls back to RSS, not PSS, so shared library pages can be counted per process: monitor_processes.py (line 22), monitor_processes.py (line 54).

3) Apple framework dependencies pulled by this libtorch build
libtorch_cpu.dylib links/weak-links Foundation, Metal/MPS, Accelerate/BLAS/LAPACK, CoreFoundation, IOKit, etc. These are mostly shared/read-only, but they still contribute to RSS accounting.

4) First tensor/linalg/autograd use
Small tensor work adds ~10 MiB in the probe. In the GP path, that corresponds to calls like torch::cdist, torch::mm, torch::linalg_cholesky, torch::cholesky_solve, and Adam setup

So if one libtorch-linked process has an extra ~80-100 MiB RSS baseline, and MPI launches multiple processes. That is why the same test shape can suddenly need more slots after being moved into the libtorch lane.

grmnptr committed:
Dont restrict PSS and AIS for libtorch only. Reduce number of seeds from 100k since that is most of the memory use.

grmnptr commented:
I decreesed the number of seeds for PSS and AIS from 100k to 1k. That will keep the tests at reasonable memory level, yielding the same results.

grmnptr committed:
Make dataio a little more elegant, reduce seeds for active learning too.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3826711/), step Results summary on 94ddd72 wanted to post the following:

## Framework test summary

Compared against 54450aa in job [civet.inl.gov/job/3821125](https://civet.inl.gov/job/3821125).

No change
## Modules test summary

Compared against 54450aa in job [civet.inl.gov/job/3821125](https://civet.inl.gov/job/3821125).

### Added tests

| Test                                                                                                | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------------------------------------|------------|---------------|
| `stochastic_tools/test:surrogates/gaussian_process.tuning/GP_squared_exponential_tuned_legacy_adam` | 2.19       | 226.46        |

### Run time changes

| Test                                                                                              | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|---------------------------------------------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `stochastic_tools/test:reporters/BFActiveLearning.sampling_bf/MultipleProc_MultipleRow_Ufunction` | 8.68       | 4.02       | -53.72% | 1290.53     | 341.87      |
| `stochastic_tools/test:reporters/AISActiveLearning.sampling/ais_al_uval`                          | 5.37       | 2.47       | -54.05% | 625.17      | 153.30      |

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3986382/), step Generate coverage on 6ff995d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4fcf7b</th>
      <th colspan=4>#30565 6ff995</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.99%</td>
      <td>86.00%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>137025</td>
      <td>137074</td>
      <td>+49</td>
      <td>58</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>22317</td>
      <td>22321</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30565/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30565/coverage/framework)


## Modules coverage

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>4fcf7b</th>
      <th colspan=4>#30565 6ff995</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.63%</td>
      <td>90.38%</td>
      <td>-0.25%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8633</td>
      <td>8728</td>
      <td>+95</td>
      <td>480</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>893</td>
      <td>929</td>
      <td>+36</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30565/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30565/coverage/stochastic_tools)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/reactor)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30565/coverage/xfem)

</details>



This comment will be updated on new commits.

grmnptr committed:
Make covariance registration conditional.

grmnptr committed:
Guard against more missing libtorch.

zachmprince commented on framework/include/libtorch/utils/LibtorchUtils.h:
```diff
@@ -25,8 +25,47 @@ namespace LibtorchUtils
+ * @param vector The vector that needs to be wrapped
+ * @param sizes The desired tensor shape
+ */
+template <typename DataType>
+torch::Tensor vectorToTensorView(std::vector<DataType> & vector, c10::IntArrayRef sizes);
```
This should probably added to LibtorchUtilsTest.C

grmnptr replied:
Great catch, will be added in the next push!

zachmprince commented on framework/include/restart/DataIO.h:
```diff
@@ -809,6 +815,10 @@ template <>
 void dataLoad(std::istream & stream, ADReal & dn, void * context);
+#ifdef MOOSE_LIBTORCH_ENABLED
+template <>
+void dataLoad(std::istream & stream, torch::Tensor & t, void * context);
```
Let's add some unit tests for the dataLoad and dataStore methods here.

grmnptr replied:
Oh yes, roger that, will be in the next push!

zachmprince commented on modules/stochastic_tools/examples/surrogates/cross_validation/all_trainers_uniform_cv_no_gp.i:
```diff
@@ -0,0 +1,142 @@
+[StochasticTools]
```
Can we just use inactive in the test spec instead of create a whole new file? I worry about having an input in the example directory that isn't referenced and is just used for testing.

grmnptr replied:
Great idea!

grmnptr replied:
The issue with this one is that Covariance syntax currently is only registered if libtorch is enabled. So we can't make it active/inactive. We could still register an empty shell to save this input file, not sure which one is a better solution. I will trust your call on this.

zachmprince replied:
Oh, I missed that the syntax is gone without libtorch. This makes sense then.

grmnptr committed:
Resolve remaining items after rebase.

zachmprince commented on modules/stochastic_tools/examples/surrogates/cross_validation/gold/all_trainers_no_libtorch.json:
```diff
@@ -21,18 +18,11 @@
-                        7.6440625444211845
+                        8.083719028212807,
+                        9.333672230328466,
+                        7.841425702031651
```
Why did these change so signficantly?

grmnptr replied:
I regolded this, but it looks like the code still gives the original results. I will reset the file.

grmnptr replied:
Well, this was not a change. We just removed the gp from this and github thinks it is smart.

zachmprince commented on modules/stochastic_tools/src/covariances/LMC.C:
```diff
@@ -149,7 +148,8 @@ LMC::computedKdhyper(RealEigenMatrix & dKdhp,
     }
 
-    MathUtils::kron(dKdhp, B, dKdhp_sub);
+    // MathUtils::kron(dKdhp, B, dKdhp_sub);
```
Remove

zachmprince commented on modules/stochastic_tools/src/libtorch/surrogates/LibtorchANNSurrogate.C:
```diff
@@ -44,12 +44,18 @@ LibtorchANNSurrogate::evaluate(const std::vector<Real> & x) const
-  mooseAssert(input_mean.size() == converted_input.size() &&
-                  input_std.size() == converted_input.size(),
+  auto input_mean_accessor = input_mean.accessor<Real, 2>();
+  auto input_std_accessor = input_std.accessor<Real, 2>();
```
What does the `2` do here? Is that the dimension? If so, shouldn't it be `1`?

grmnptr replied:
The standardizer stores these in a [n, 1] tensor. Why not just [n]? That is a great question, I will modify this.

zachmprince commented on modules/stochastic_tools/src/surrogates/ActiveLearningGaussianProcess.C:
```diff
@@ -79,29 +90,37 @@ ActiveLearningGaussianProcess::reTrain(const std::vector<std::vector<Real>> & in
 
+  _training_params =
+      LibtorchUtils::vectorToTensorCopy(flat_inputs, {long(outputs.size()), long(input_size)});
+  _training_data = LibtorchUtils::vectorToTensorCopy(outputs, {long(outputs.size()), 1});
```
I'm curious why it's necessary to have the `long` conversion. Does that not happen automatically?

grmnptr replied:
Well, I was thinking: size_t is unsigned so we could run into issues where it goes over the signed representation limits.  But I suspect the libmesh convert int functions could be better here. Will add that!

zachmprince commented on modules/stochastic_tools/src/utils/GaussianProcess.C:
```diff
@@ -67,35 +150,49 @@ GaussianProcess::linkCovarianceFunction(CovarianceFunctionBase * covariance_func
+  setupStoredMatrices(flattened_tensor);
+
+  _hyperparam_map.clear();
+  _covariance_function->buildHyperParamMap(_hyperparam_map);
```
This seems duplicative to line 173-174

grmnptr replied:
Well, if you are training then:
We build the map -> Convert it to double/move it to device -> Reload it into the covariance objects
then we train (which loads the updated values into the covariance)
Then we rebuild to make sure we have the latest map.

I think the last one is not needed because the tune function already updates the map within the covariance objects.

Will test this guy.

grmnptr replied:
You were right, these are not needed. Will remove them.

zachmprince commented on modules/stochastic_tools/src/utils/GaussianProcess.C:
```diff
@@ -133,136 +230,163 @@ GaussianProcess::generateTuningMap(const std::vector<std::string> & params_to_tu
+  const auto options = doubleOptionsLike(training_params);
+  std::vector<Real> theta_values(_num_tunable, 0.0);
+  _hyperparam_map.clear();
+  _covariance_function->buildHyperParamMap(_hyperparam_map);
```
And again here?

grmnptr replied:
Removed

zachmprince commented on modules/stochastic_tools/src/actions/LoadCovarianceDataAction.C:
```diff
@@ -7,13 +7,67 @@
+      mooseError("Expected vector hyperparameter for ", param_name, ".");
+    params.set<std::vector<Real>>(param_name) = exportHyperParameter(tensor);
+  }
+}
```
Let's add an error if the parameter is of unsupported type.

grmnptr replied:
Will do!

zachmprince commented on framework/src/restart/DataIO.C:
```diff
@@ -555,6 +582,34 @@ dataLoad(std::istream & stream, std::stringstream & s, void * /* context */)
+#ifdef MOOSE_LIBTORCH_ENABLED
+template <>
+void
+dataLoad(std::istream & stream, torch::Tensor & t, void * context)
```
Do we need to make sure to load the tensor on the same device it was stored on?

grmnptr replied:
that is a great question, I never thought about it. I suppose if we run with the same device setting then it is probably doable. Let me check a few things.

grmnptr replied:
I mean right now we are moving everything to host in the save and loading back onto the host. I think that is a safe option for now. I suppose we could do it, but I am not sure how big the differences are in terms of device and device and what we need to store to be able to make sure we are loading back oxactly the same thing. Because device number might not be enough on its own.

grmnptr replied:
If it is okay with you, I wouldn't do it in this PR.

zachmprince replied:
I think this is fine. The store technique is copying to host, so load should follow the same pattern. We will probably have to think carefully when we start playing with recover. Which I'm doing in #33295

zachmprince commented on modules/stochastic_tools/test/tests/reporters/ActiveLearningGP/tests:
```diff
@@ -9,6 +9,9 @@
       detail = 'a single row in the sampler using the U learning function;'
+      capabilities = 'libtorch'
+      compute_devices = 'cpu cuda'
+      min_slots = 3 # Reserve extra CI memory for libtorch runtime in GP tests
```
Can you check whether this is still necessary?

grmnptr replied:
Definitely hitting these with regular settings.
I got these last time and nothing changed since then. I actually tried to use the least amount of slots.
```
JOB KILLED (OVER MEMORY): Memory/slot 311.96 MB > allowed 300.00 MB
```
We might have to increase the allowed limits in STM if we want to reduce the slots in the tests.

zachmprince replied:
That limit is useful to see if framework changes cause memory spikes. If we need this many slots for this test now, then we need it. I was mainly just surprised, even with reducing the number of random number seeds.

zachmprince commented on modules/stochastic_tools/test/tests/reporters/BFActiveLearning/gold/multiple_proc_multiple_row_ufunction.json:
```diff
@@ -1230,14 +1230,122 @@
+                ]
+            },
+            "time": 23.0,
+            "time_step": 23
```
I wanted to bring this re-gold to @somu15's attention. The std toward the end seem pretty different, causing the extra time steps.

zachmprince reviewed (approved):
Great work @DanielYankura  and @grmnptr! I don't think I have anything big here, mainly clean-up and some testing suggestions.

zachmprince reviewed (changes requested):
Dismissing my erroneous approval above

grmnptr committed:
Add unit test for the tensor view function.

grmnptr committed:
Add data IO unit test for torch tensors.

grmnptr committed:
Implement changes in the standardizer, propagate to the ANN surrogate.

grmnptr committed:
Clean up unnecessary map builds.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3985042/), step Clang format on 9a0cc68 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30565/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30565/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4fcf7b8445ffc056cb022a336c473524e2d4d1b6`

grmnptr committed:
Remove gp from nolibtorch runs.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3986411/), step Results summary on 6ff995d wanted to post the following:

## Framework test summary

Compared against 4fcf7b8 in job [civet.inl.gov/job/3980633](https://civet.inl.gov/job/3980633).

No change
## Modules test summary

Compared against 4fcf7b8 in job [civet.inl.gov/job/3980633](https://civet.inl.gov/job/3980633).

### Added tests

| Test                                                                                                | Time (s)   | Memory (MB)   |
|-----------------------------------------------------------------------------------------------------|------------|---------------|
| `stochastic_tools/test:surrogates/gaussian_process.tuning/GP_squared_exponential_tuned_legacy_adam` | 2.05       | 237.89        |

grmnptr commented:
Thanks @DanielYankura for the great work and thanks @zachmprince for the great review!
